ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatAgent.java

```diff
@@ -0,0 +1,496 @@
+   *
+   * <p>This is an asynchronous RPC and we do not handle the response returned from the remote
+   * server.</p>
+   */
+  class SendHeartbeatService extends AbstractScheduledService {
+
+    @Override
+    protected void runOneIteration() {
+      for (Entry<String, HostStatusEntity> hostStatusEntry: hostsStatus.entrySet()) {
+        final String host = hostStatusEntry.getKey();
+        final HostStatusEntity status = hostStatusEntry.getValue();
+        try {
+          if (!host.equals(localHostString)) {
+            final URI remoteUri = buildLocation(localURL, status.getHostInfoEntity().getHost(),
+                                          status.getHostInfoEntity().getPort());
+            log.debug("Send heartbeat to host {} at {}", status.getHostInfoEntity().getHost(),
+                      clock.millis());
+            serviceContext.getKsqlClient().makeAsyncHeartbeatRequest(remoteUri, localHostInfo,
+                                                                     clock.millis());
+          }
+        } catch (Throwable t) {
+          log.error("Request to server: " + status.getHostInfoEntity().getHost() + ":"
+                        + status.getHostInfoEntity().getPort()
+                        + " failed with exception: " + t.getMessage(), t);
+        }
+      }
+    }
+
+    @Override
+    protected Scheduler scheduler() {
+      return Scheduler.newFixedRateSchedule(SEND_HEARTBEAT_DELAY_MS,
+                                            config.heartbeatSendIntervalMs,
+                                            TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    protected ScheduledExecutorService executor() {
+      return scheduledExecutorService;
+    }
+
+    private URI buildLocation(final URL localHost, final String host, final int port) {
+      try {
+        return new URL(localHost.getProtocol(), host, port, "/").toURI();
+      } catch (final Exception e) {
+        throw new IllegalStateException("Failed to convert remote host info to URL."
+                                            + " remoteInfo: " + host + ":" + port);
+      }
+    }
+  }
+
+  /**
+   * Discovers remote hosts in the cluster through the metadata of currently running
+   * persistent queries.
+   */
+  class DiscoverClusterService extends AbstractScheduledService {
+
+    @Override
+    protected void runOneIteration() {
+      try {
+        final List<PersistentQueryMetadata> currentQueries = engine.getPersistentQueries();
+        if (currentQueries.isEmpty()) {
+          return;
+        }
+
+        final Set<HostInfo> uniqueHosts = currentQueries.stream()
+            .map(queryMetadata -> ((QueryMetadata) queryMetadata).getAllMetadata())
+            .filter(Objects::nonNull)
+            .flatMap(Collection::stream)
+            .map(StreamsMetadata::hostInfo)
+            .filter(hostInfo -> !(hostInfo.host().equals(localHostInfo.host())
+                && hostInfo.port() == (localHostInfo.port())))
```

Comment:
why do the other hosts need to have the same heartbeat port?

Reply:
They don't. I am filtering out the hosts that match the localhost. The `!` is around both conditions

Reply:
derp!